pyfixest/estimation/feiv_.py

```diff
@@ -193,3 +206,263 @@ def get_fit(self) -> None:
+            print("Effective F stat :", F_stat_eff_pf)
+
+            # The example above generates the following results
+            # (Unadjusted) F stat : 52.81535560457482
+            # Effective F stat : 48.661542741328205
+        """
+        iv_diag_statistics = iv_diag_statistics or []
+
+        if "f_stat" in iv_diag_statistics:
+            self._p_iv = len(self._non_exo_instruments)
+
+            # Create an identity matrix of size p_iv by p_iv
+            # Pad the identity matrix with zeros to make it of size p_iv by k
+            p_iv = self._p_iv  # number of IVs
+            k = (
+                self._model_1st_stage._k
+            )  # number of estimated coefficients of 1st stage
+
+            # Generate matrix R that tests the following;
+            # H0 : \beta_{z_1} = 0 & ... & \beta_{z_{p_iv}} = 0
+            #      where z_1, ..., z_{p_iv} are the instrument variables
+            # H1 : H0 does not hold
+
+            identity_matrix = np.eye(p_iv)
+            R = np.zeros((p_iv, k))
+            if self._has_fixef:
+                R[:, :p_iv] = identity_matrix
+            else:
+                R[:, 1 : p_iv + 1] = identity_matrix
+
+            self._model_1st_stage.wald_test(R=R)
+            self._f_stat_1st_stage = self._model_1st_stage._f_statistic
+            self._p_value_1st_stage = self._model_1st_stage._p_value
+
+        if "effective_f" in iv_diag_statistics:
+            self.eff_F()
+
+    def eff_F(self) -> None:
+        """Compute Effective F stat (Olea and Pflueger 2013)."""
+        # If vcov is iid, redo first stage regression
+        if self._vcov_type_detail == "iid":
+            self._vcov_type_detail = "hetero"
+            self.first_stage()
+
+        # Compute Effective F stat by Olea and Pflueger 2013
+        # 1. Extract First Stage Coefficients and Variance-Covariance Matrix:
+        #   Extract the coefficients for the instrument z
+        #   from the first stage regression.
+        #   Extract the robust variance-covariance matrix of these coefficients.
+        #   Extract the instrument matrix.
+        # 2. Compute the Instrument Matrix:
+        #   Construct the instrument matrix Q_{zz} = Z.T x Z
+        # 3. Compute the Effective F-statistic:
+        #   F_{eff} = π.T Q_{zz} π / trance(ΣQ_{zz})
+
+        # Extract coefficients for the non-exogenous instruments
+
+        pi_hat = np.array(
+            [
+                self._model_1st_stage.coef()[instrument]
+                for instrument in self._non_exo_instruments
+            ]
+        )
+
+        Z = self._Z_iv
+
+        # Calculate the cross-product of the instrument matrix
+        Q_zz = Z.T @ Z
+
+        # Extract the robust variance-covariance matrix
+        vcv = self._model_1st_stage._vcov
+
+        # Map the instrument names to their indices in the parameter list
+        # Number of rows/columns in vcv
+
+        # Extract the submatrix
```

Comment:
Here again it might be "safer" to extract the position of the instruments based on "explicit" knowledge rather than "implicit" knowledge"?

Reply:
Actually, you pretty much do this here I think 😄 

```python
            [
                self._model_1st_stage.coef()[instrument]
                for instrument in self._non_exo_instruments
            ]
```